haystack/dataclasses/chat_message.py

```diff
@@ -5,104 +5,306 @@
+        Convert a string to a ChatRole enum.
+        """
+        enum_map = {e.value: e for e in ChatRole}
+        role = enum_map.get(string)
+        if role is None:
+            msg = f"Unknown chat role '{string}'. Supported roles are: {list(enum_map.keys())}"
+            raise ValueError(msg)
+        return role
+
+
+@dataclass
+class ToolCall:
+    """
+    Represents a Tool call prepared by the model, usually contained in an assistant message.
+
+    :param id: The ID of the Tool call.
+    :param tool_name: The name of the Tool to call.
+    :param arguments: The arguments to call the Tool with.
+    """
+
+    tool_name: str
+    arguments: Dict[str, Any]
+    id: Optional[str] = None  # noqa: A003
+
+
+@dataclass
+class ToolCallResult:
+    """
+    Represents the result of a Tool invocation.
+
+    :param result: The result of the Tool invocation.
+    :param origin: The Tool call that produced this result.
+    :param error: Whether the Tool invocation resulted in an error.
+    """
+
+    result: str
+    origin: ToolCall
+    error: bool
+
+
+@dataclass
+class TextContent:
+    """
+    The textual content of a chat message.
+
+    :param text: The text content of the message.
+    """
+
+    text: str
+
+
+ChatMessageContentT = Union[TextContent, ToolCall, ToolCallResult]
+
 
 @dataclass
 class ChatMessage:
     """
     Represents a message in a LLM chat conversation.
 
-    :param content: The text content of the message.
-    :param role: The role of the entity sending the message.
-    :param name: The name of the function being called (only applicable for role FUNCTION).
-    :param meta: Additional metadata associated with the message.
+    Use the `from_assistant`, `from_user`, `from_system`, and `from_tool` class methods to create a ChatMessage.
     """
 
-    content: str
-    role: ChatRole
-    name: Optional[str]
-    meta: Dict[str, Any] = field(default_factory=dict, hash=False)
+    _role: ChatRole
+    _content: Sequence[ChatMessageContentT]
+    _meta: Dict[str, Any] = field(default_factory=dict, hash=False)
 
-    @property
-    def text(self) -> Optional[str]:
+    def __new__(cls, *args, **kwargs):
         """
-        Returns the textual content of the message.
+        This method is reimplemented to make the changes to the `ChatMessage` dataclass more visible.
         """
-        # Currently, this property mirrors the `content` attribute. This will change in 2.9.0.
-        # The current actual return type is str. We are using Optional[str] to be ready for 2.9.0,
-        # when None will be a valid value for `text`.
-        return object.__getattribute__(self, "content")
+
+        general_msg = (
+            "Use the `from_assistant`, `from_user`, `from_system`, and `from_tool` class methods to create a "
+            "ChatMessage. For more information about the new API and how to migrate, see the documentation:"
+            " https://docs.haystack.deepset.ai/docs/data-classes#chatmessage"
```

Comment:
this link and the following ones will be updated once we have a new documentation page containing examples and the migration guide